### PR TITLE
refactor(iot-service, prov-service): Remove TlsVersions

### PR DESCRIPTION
**Pull request by brycewang-microsoft**
+ **IoT Hub service client**: 
    * remove "TlsVersions" and move properties into "IotHubServiceClientOptions".
    *We usually put TlsVersions properties into the transport settings classes. But APIs in service clients are using different protocols, and we are using "IotHubServiceClientOptions" to keep other properties as well (i.e., Proxy and HttpClient).*

<br>

+ **Provisioning service client**:
    + add "ProvisioningServiceClientOptions" and pass it into the constructor of "ProvisioningServiceClient".
    + rename "HttpTransportSettings" to "ProvisioningServiceHttpSettings" to align with naming format everywhere else.
    + remove "TlsVersions" and move properties into "ProvisioningServiceHttpSettings".

### CODE REVIEW

File: common/src/service/HttpClientHelper.cs [hunk 1]
```diff
@@ -871,8 +871,9 @@ internal static HttpClient CreateDefaultClient(IWebProxy webProxy, Uri baseUri,
         internal static HttpMessageHandler CreateDefaultHttpMessageHandler(IWebProxy webProxy, Uri baseUri, int connectionLeaseTimeoutMilliseconds)
         {
             var httpMessageHandler = new HttpClientHandler();
-            httpMessageHandler.SslProtocols = TlsVersions.Instance.Preferred;
-            httpMessageHandler.CheckCertificateRevocationList = TlsVersions.Instance.CertificateRevocationCheck;
+            var options = new IotHubServiceClientOptions();
+            httpMessageHandler.SslProtocols = options.SslProtocols;
```
Comment (drwill-ms) [inline common/src/service/HttpClientHelper.cs:875]: Need to pipe in the IotHubServiceClientOptions that the user passed in the service client ctor.

File: iothub/service/src/IotHubClientWebSocket.cs [hunk 2]
```diff
@@ -144,7 +144,7 @@ await sslStream
                         .AuthenticateAsClientAsync(
                             host,
                             x509CertificateCollection,
-                            TlsVersions.Instance.Preferred,
+                            new IotHubServiceClientOptions().SslProtocols,
```
Comment (drwill-ms) [inline iothub/service/src/IotHubClientWebSocket.cs:147]: Need to get this from the customer from the service client ctor.

File: provisioning/service/src/ProvisioningServiceClient.cs [hunk 3]
```diff
@@ -77,32 +77,20 @@ public class ProvisioningServiceClient : IDisposable
...
         /// <exception cref="ArgumentException">If the connectionString is <c>null</c> or empty.</exception>
-        public static ProvisioningServiceClient CreateFromConnectionString(string connectionString)
+        public static ProvisioningServiceClient CreateFromConnectionString(string connectionString, ProvisioningServiceClientOptions options = default)
         {
-            return new ProvisioningServiceClient(connectionString, new HttpTransportSettings());
-        }
+            if (options == null)
+            {
+                options = new ProvisioningServiceClientOptions();
+            }
```
Comment (drwill-ms) [inline provisioning/service/src/ProvisioningServiceClient.cs:88]:
```suggestion
            if (options == default)
            {
                options = new();
            }
```

File: provisioning/service/src/ProvisioningServiceClient.cs [hunk 4]
```diff
@@ -449,7 +437,7 @@ public Query CreateIndividualEnrollmentQuery(QuerySpecification querySpecificati
             return IndividualEnrollmentManager.CreateQuery(
                 _provisioningConnectionString,
                 querySpecification,
-                new HttpTransportSettings(),
+                new ProvisioningServiceHttpSettings(),
```
Comment (drwill-ms) [inline provisioning/service/src/ProvisioningServiceClient.cs:440]: Should we be getting this from `options`?

File: provisioning/service/src/ProvisioningServiceClientOptions.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        /// </summary>
+        public ProvisioningServiceClientOptions()
+        {
+            ProvisioningServiceHttpSettings = new ProvisioningServiceHttpSettings();
+        }
+
+        /// <summary>
+        /// The <see cref="ProvisioningServiceHttpSettings"/> transport settings to use.
+        /// </summary>
+        public ProvisioningServiceHttpSettings ProvisioningServiceHttpSettings { get; }
```
Comment (drwill-ms) [inline provisioning/service/src/ProvisioningServiceClientOptions.cs:23]:
```suggestion
        /// <summary>
        /// The transport settings to use.
        /// </summary>
        public ProvisioningServiceHttpSettings ProvisioningServiceHttpSettings { get; } = new();
```

File: provisioning/service/src/ProvisioningServiceHttpSettings.cs [hunk 6]
```diff
@@ -2,18 +2,19 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System.Net;
+using System.Security.Authentication;
 
 namespace Microsoft.Azure.Devices.Provisioning.Service
 {
     /// <summary>
     /// contains Http1 transport-specific settings for Service Client
```
Comment (drwill-ms) [inline provisioning/service/src/ProvisioningServiceHttpSettings.cs:10]:
```suggestion
    /// Contains HTTP transport-specific settings for the provisioning service client
```

File: provisioning/service/src/ProvisioningServiceHttpSettings.cs [hunk 7]
```diff
@@ -22,5 +23,18 @@ public HttpTransportSettings()
         /// Gets or sets proxy information for the request.
         /// </summary>
         public IWebProxy Proxy { get; set; }
```
Comment (drwill-ms) [inline provisioning/service/src/ProvisioningServiceHttpSettings.cs:25]:
```suggestion
        public IWebProxy Proxy { get; set; } = DefaultWebProxySettings.Instance;
```

Comment (drwill-ms) [reply provisioning/service/src/ProvisioningServiceHttpSettings.cs:25]: Then remove ctor.

File: provisioning/service/src/ProvisioningServiceHttpSettings.cs [hunk 8]
```diff
@@ -22,5 +23,18 @@ public HttpTransportSettings()
...
+        /// The version of TLS to use by default.
+        /// </summary>
+        /// <remarks>
+        /// Defaults to "None", which means let the OS decide the proper TLS version (SChannel in Windows / OpenSSL in Linux).
+        /// </remarks>
+        public SslProtocols SslProtocols { get; set; } = SslProtocols.None;
+
+        /// <summary>
+        /// To enable certificate revocation check. Default to be false.
+        /// </summary>
```
Comment (drwill-ms) [inline provisioning/service/src/ProvisioningServiceHttpSettings.cs:37]:
```suggestion
        /// To enable certificate revocation check.
        /// </summary>
        /// <remarks>
        /// Defaults to false.
        /// </remarks>
```

Comment (drwill-ms) [reply provisioning/service/src/ProvisioningServiceHttpSettings.cs:37]: Also earlier in this PR.

File: common/src/service/HttpClientHelper.cs [hunk 9]
```diff
@@ -871,8 +871,9 @@ internal static HttpClient CreateDefaultClient(IWebProxy webProxy, Uri baseUri,
         internal static HttpMessageHandler CreateDefaultHttpMessageHandler(IWebProxy webProxy, Uri baseUri, int connectionLeaseTimeoutMilliseconds)
         {
             var httpMessageHandler = new HttpClientHandler();
-            httpMessageHandler.SslProtocols = TlsVersions.Instance.Preferred;
-            httpMessageHandler.CheckCertificateRevocationList = TlsVersions.Instance.CertificateRevocationCheck;
+            var options = new IotHubServiceClientOptions();
```
Comment (brycewang-microsoft) [inline common/src/service/HttpClientHelper.cs:874]: This class is about to remove

**Comment by brycewang-microsoft** [comment]
azp run

**Review by drwill-ms — APPROVED** [review]

**Pull request merged**